README
NAME
    pzem is a PZEM004T power meter client

AUTHORS
    Viacheslav Chimishuk <vchimishuk@yandex.ru>

COPYING
    This programm is released under the GNU General Public License version 3 or later, which is
    distributed in the COPYING file. You should have received a copy of the GNU General Public License along with
    this program.  If not, see <http://www.gnu.org/licenses/>.

USAGE
    $ pzem --stats
    voltage: 215.2
    current: 0.129
    power: 15.1
    energy: 1
    frequency: 50.0
    factor: 0.54
    alarm: false

    Current distribution also includes Python package.
    Python package usage example:
    >>> import pzem
    >>> p = pzem.Pzem('/dev/ttyUSB0')
    >>> p.stats()
    >>> {'voltage': 215.2, 'current': 0.129, 'power': 15.1, 'energy': 1,
    ... 'frequency': 50.0, 'factor': 0.54, 'alarm': False}


debian/changelog
pzem (0.1.0) stable; urgency=low

  * Initial release.

 -- Viacheslav Chimishuk <vchimishuk@yandex.ru>  Sun, 12 Mar 2023 19:40:00 +0200


pzem/crc.py
TABLE = [0X0000, 0XC0C1, 0XC181, 0X0140, 0XC301, 0X03C0, 0X0280, 0XC241,
         0XC601, 0X06C0, 0X0780, 0XC741, 0X0500, 0XC5C1, 0XC481, 0X0440,
         0XCC01, 0X0CC0, 0X0D80, 0XCD41, 0X0F00, 0XCFC1, 0XCE81, 0X0E40,
         0X0A00, 0XCAC1, 0XCB81, 0X0B40, 0XC901, 0X09C0, 0X0880, 0XC841,
         0XD801, 0X18C0, 0X1980, 0XD941, 0X1B00, 0XDBC1, 0XDA81, 0X1A40,
         0X1E00, 0XDEC1, 0XDF81, 0X1F40, 0XDD01, 0X1DC0, 0X1C80, 0XDC41,
         0X1400, 0XD4C1, 0XD581, 0X1540, 0XD701, 0X17C0, 0X1680, 0XD641,
         0XD201, 0X12C0, 0X1380, 0XD341, 0X1100, 0XD1C1, 0XD081, 0X1040,
         0XF001, 0X30C0, 0X3180, 0XF141, 0X3300, 0XF3C1, 0XF281, 0X3240,
         0X3600, 0XF6C1, 0XF781, 0X3740, 0XF501, 0X35C0, 0X3480, 0XF441,
         0X3C00, 0XFCC1, 0XFD81, 0X3D40, 0XFF01, 0X3FC0, 0X3E80, 0XFE41,
         0XFA01, 0X3AC0, 0X3B80, 0XFB41, 0X3900, 0XF9C1, 0XF881, 0X3840,
         0X2800, 0XE8C1, 0XE981, 0X2940, 0XEB01, 0X2BC0, 0X2A80, 0XEA41,
         0XEE01, 0X2EC0, 0X2F80, 0XEF41, 0X2D00, 0XEDC1, 0XEC81, 0X2C40,
         0XE401, 0X24C0, 0X2580, 0XE541, 0X2700, 0XE7C1, 0XE681, 0X2640,
         0X2200, 0XE2C1, 0XE381, 0X2340, 0XE101, 0X21C0, 0X2080, 0XE041,
         0XA001, 0X60C0, 0X6180, 0XA141, 0X6300, 0XA3C1, 0XA281, 0X6240,
         0X6600, 0XA6C1, 0XA781, 0X6740, 0XA501, 0X65C0, 0X6480, 0XA441,
         0X6C00, 0XACC1, 0XAD81, 0X6D40, 0XAF01, 0X6FC0, 0X6E80, 0XAE41,
         0XAA01, 0X6AC0, 0X6B80, 0XAB41, 0X6900, 0XA9C1, 0XA881, 0X6840,
         0X7800, 0XB8C1, 0XB981, 0X7940, 0XBB01, 0X7BC0, 0X7A80, 0XBA41,
         0XBE01, 0X7EC0, 0X7F80, 0XBF41, 0X7D00, 0XBDC1, 0XBC81, 0X7C40,
         0XB401, 0X74C0, 0X7580, 0XB541, 0X7700, 0XB7C1, 0XB681, 0X7640,
         0X7200, 0XB2C1, 0XB381, 0X7340, 0XB101, 0X71C0, 0X7080, 0XB041,
         0X5000, 0X90C1, 0X9181, 0X5140, 0X9301, 0X53C0, 0X5280, 0X9241,
         0X9601, 0X56C0, 0X5780, 0X9741, 0X5500, 0X95C1, 0X9481, 0X5440,
         0X9C01, 0X5CC0, 0X5D80, 0X9D41, 0X5F00, 0X9FC1, 0X9E81, 0X5E40,
         0X5A00, 0X9AC1, 0X9B81, 0X5B40, 0X9901, 0X59C0, 0X5880, 0X9841,
         0X8801, 0X48C0, 0X4980, 0X8941, 0X4B00, 0X8BC1, 0X8A81, 0X4A40,
         0X4E00, 0X8EC1, 0X8F81, 0X4F40, 0X8D01, 0X4DC0, 0X4C80, 0X8C41,
         0X4400, 0X84C1, 0X8581, 0X4540, 0X8701, 0X47C0, 0X4680, 0X8641,
         0X8201, 0X42C0, 0X4380, 0X8341, 0X4100, 0X81C1, 0X8081, 0X4040]


# CRC16 implementation is based on https://www.modbustools.com/modbus_crc16.html
def crc16(buf):
    crc = 0xFFFF

    for b in buf:
        t = (b ^ crc) & 0xFF
        crc >>= 8
        crc ^= TABLE[t]

    return crc


pzem/pzem.py
import serial
from . import crc


class Pzem:
    ADDR_DEFAULT = 0xF8
    # Read Input Registers
    CMD_RIR = 0x04
    CMD_RIR_ERR = 0x84
    # Reset.
    CMD_RST = 0x42
    CMD_RST_ERR = 0xC2
    ERR_CODE = {0x01: 'illegal function',
                0x02: 'illegal address',
                0x03: 'illegal data',
                0x04: 'slave error'}

    def __init__(self, dev, addr=ADDR_DEFAULT):
        self.addr = addr
        self.ser = serial.Serial(port=dev, baudrate=9600,
                                 bytesize=serial.EIGHTBITS,
                                 parity=serial.PARITY_NONE,
                                 stopbits=serial.STOPBITS_ONE,
                                 timeout=3)

    def stats(self):
        def read_body():
            n = self.readn(1)[0]
            return self.readn(n)

        self.write([self.addr, Pzem.CMD_RIR, 0x00, 0x00, 0x00, 0x0A])
        data = self.read(Pzem.CMD_RIR, Pzem.CMD_RIR_ERR, read_body)

        return {'voltage': int.from_bytes(data[0:2],
                                          byteorder='big') / 10,
                'current': int.from_bytes(data[4:6] + data[2:4],
                                          byteorder='big') / 1000,
                'power': int.from_bytes(data[8:10] + data[6:8],
                                        byteorder='big') / 10,
                'energy': int.from_bytes(data[12:14] + data[10:12],
                                         byteorder='big'),
                'frequency': int.from_bytes(data[14:16],
                                            byteorder='big') / 10,
                'factor': int.from_bytes(data[16:18],
                                         byteorder='big') / 100,
                'alarm': int.from_bytes(data[18:20],
                                        byteorder='big') == 0xFFFF}

    def reset(self):
        def read_body():
            pass

        self.write([self.addr, Pzem.CMD_RST])
        self.read(Pzem.CMD_RST, Pzem.CMD_RST_ERR, read_body)

    def close(self):
        self.ser.close()

    def read(self, cmd, err, read_body):
        addr = self.readn(1)[0]
        if addr != self.addr:
            raise IOError(0, 'expected response for addr {}, got {}'
                          .format(self.addr, addr))

        rcmd = self.readn(1)[0]
        if rcmd != cmd:
            if rcmd == err:
                e = self.readn(1)[0]
                # Consume CRC.
                self.readn(2)
                raise IOError(e, Pzem.ERR_CODE.get(e, 'unknown error'))
            else:
                raise IOError(rcmd,
                              Pzem.ERR_CODE.get(e, 'invalid response'))

        b = read_body()
        # Consume CRC.
        self.readn(2)

        return b

    def write(self, buf):
        c = crc.crc16(buf)
        buf += [c & 0xFF]
        buf += [(c >> 8) & 0xFF]
        buflen = len(buf)

        n = self.ser.write(bytes(buf))
        if n != buflen:
            raise IOError(0, 'wrote only {} of {}'.format(n, buflen))

    def readn(self, n):
        b = self.ser.read(n)
        if len(b) == 0:
            raise IOError(0, 'read timed out')

        return b
